### PR TITLE
Create credit memo for sales invoice

### PR DESCRIPTION
**Pull request by TheBestPessimist**
https://github.com/metasfresh/metasfresh-e2e/issues/133

### CODE REVIEW

File: cypress/fixtures/misc/misc_dictionary.json [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
+  "docStatusCompleted": "Completed",
+  "de_DE__docStatusCompleted": "Fertiggestellt",
 
-    "defaultTaxCategory_comment": "used in tax tests and also in places where we just need to set some tax rate",
-    "defaultTaxCategory": "Regular Tax Rate 19% (Germany)",
-    "de_DE__defaultTaxCategory": "Normaler Steuersatz 19% (Deutschland)"
-}
\ No newline at end of file
+  "docStatusInProgress": "In Progress",
+  "de_DE__docStatusInProgress": "In Progress",
```
Comment (metas-ts) [inline cypress/fixtures/misc/misc_dictionary.json]:
```suggestion
  "de_DE__docStatusInProgress": "In Verarbeitung",
```

File: cypress/integration/sales/credit_memo_for_sales_invoice.js [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        .setDocumentAction(getLanguageSpecific(salesInvoiceJson, 'docActionComplete'))
+        .setDocumentStatus(getLanguageSpecific(salesInvoiceJson, 'Completed'))
+        .apply();
+    });
+
+    cy.readAllNotifications();
+  });
+
+
+  // it('Open Sales Invoice Window', () => {
```
Comment (metas-ts) [inline cypress/integration/sales/credit_memo_for_sales_invoice.js]: pls remove this

File: cypress/fixtures/misc/misc_dictionary.json (hunk 1 above)
Comment (metas-ts) [inline cypress/fixtures/misc/misc_dictionary.json]:
```suggestion
  "de_DE__docStatusInProgress": "In Verarbeitung",
```

..C&Ped this trl from the application ditionary, so we can be confident about it

File: cypress/integration/sales/credit_memo_for_sales_invoice.js [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+  it('Open the Referenced Sales Invoice documents', function () {
+    cy.wait(5000);  // give the system some breathing space
+    cy.get('body').type('{alt}6'); // open referenced-records-sidelist
+    cy.selectReference('AD_RelationType_ID-540184').click();
+  });
+
+
+  it('Ensure there is only 1 Sales Invoice row and open it', function () {
+    // salesInvoices.getRows().should('have.length', 1);
```
Comment (metas-ts) [inline cypress/integration/sales/credit_memo_for_sales_invoice.js]: pls comment back in ro remove
...or add another comment about why is should be like this

File: cypress/support/commands/byTbp.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    let currentRecordId = 0;
+    cy.url().then(ulrr => {
+      currentRecordId = ulrr.split('/').pop();
+    });
+    return currentRecordId;
+  });
+});
+
+
+Cypress.Commands.add('setCheckBoxValue', (fieldName, isChecked, modal = false, rewriteUrl = null) => {
```
Comment (metas-ts) [inline cypress/support/commands/byTbp.js]: i believe this one should go to `form.js`

File: cypress/support/commands/byTbp.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
+/*
+@kuba idk where all these should be placed so i just threw them here.
+
+Pls tell me the _organisation_ standards for cypress/js :^).
+ */
+
+
+import {getLanguageSpecific} from "../utils/utils";
+
+Cypress.Commands.add('getCurrentRecordId', () => {
```
Comment (metas-ts) [inline cypress/support/commands/byTbp.js]: maybe this one can go to `general.js`?

File: cypress/support/commands/byTbp.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  /**
+   * WINDOW is the default
+   */
+  static WINDOW = '/rest/api/window/.*[^/][^N][^E][^W]$';
+
+  static PROCESS = '/rest/api/process/';
+}
+
+
+Cypress.Commands.add('expectDocumentStatus', (expectedDocumentStatus) => {
```
Comment (metas-ts) [inline cypress/support/commands/byTbp.js]: maybe this one can to to `test.js`..

File: cypress/support/commands/byTbp.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        }
+      }
+    });
+  });
+});
+
+/**
+ * These constants should be used instead of writing strings every time
+ */
+export class RewriteURL {
```
Comment (metas-ts) [inline cypress/support/commands/byTbp.js]: maybe we can find a new, dedicated place for this one, and also for `DocumentStatusKey` ?

File: cypress/integration/sales/credit_memo_for_sales_invoice.js (hunk 3 above)
Comment (TheBestPessimist) [reply cypress/integration/sales/credit_memo_for_sales_invoice.js]: uncommented.

File: cypress/integration/sales/credit_memo_for_sales_invoice.js (hunk 2 above)
Comment (TheBestPessimist) [reply cypress/integration/sales/credit_memo_for_sales_invoice.js]: done

File: cypress/support/commands/byTbp.js (hunk 5 above)
Comment (siemiatj) [reply cypress/support/commands/byTbp.js]: Yup, should be good

File: cypress/support/commands/byTbp.js (hunk 4 above)
Comment (siemiatj) [reply cypress/support/commands/byTbp.js]: why is `clickOnCheckBox` not enough ?

File: cypress/support/commands/byTbp.js (hunk 7 above)
Comment (siemiatj) [reply cypress/support/commands/byTbp.js]: Yeah, I feel like this should land in functions.

File: cypress/support/commands/byTbp.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  describe(`Expect specific document status`, function () {
+    cy.fixture('misc/misc_dictionary.json').then(miscDictionaryJson => {
+      const expectedTrl = getLanguageSpecific(miscDictionaryJson, expectedDocumentStatus);
+      const documentTag = DocumentStatusKey[`_tag_${expectedDocumentStatus}`];
+      cy.get(`.meta-dropdown-toggle ${documentTag}`).contains(expectedTrl);
+    });
+  });
+});
+
+export class DocumentStatusKey {
```
Comment (siemiatj) [inline cypress/support/commands/byTbp.js]: maybe we need some constants file as well ?

File: cypress/support/commands/byTbp.js (hunk 4 above)
Comment (TheBestPessimist) [reply cypress/support/commands/byTbp.js]: done

File: cypress/support/commands/byTbp.js (hunk 5 above)
Comment (TheBestPessimist) [reply cypress/support/commands/byTbp.js]: done

File: cypress/support/commands/byTbp.js (hunk 6 above)
Comment (TheBestPessimist) [reply cypress/support/commands/byTbp.js]: done

File: cypress/support/commands/byTbp.js (hunk 8 above)
Comment (TheBestPessimist) [reply cypress/support/commands/byTbp.js]: added `cypress/support/utils/constants.js`

File: cypress/support/commands/byTbp.js (hunk 7 above)
Comment (TheBestPessimist) [reply cypress/support/commands/byTbp.js]: added `cypress/support/utils/constants.js`

**Pull request merged**